### PR TITLE
WIP: Platform nir0s distro

### LINKED ISSUE #30489: Remove use of platform.dist() and platform.linux_distribution() methods removed in python 3.7
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
module_utils/basic.py
assorted others

##### ANSIBLE VERSION
```
ansible 2.5.0 (issue/23485 f4245e31f0) last updated 2017/09/17 18:18:48 (GMT -500)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/reid/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/reid/git/ansible/lib/ansible
  executable location = /home/reid/git/ansible/bin/ansible
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]
```

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
N/A

##### SUMMARY
Several scripts use deprecated platform.dist() method. Expected to be removed in Py3.7 (https://bugs.python.org/issue1322). No plans for a standard library replacement, but nir0s/distro aims to serve the same purpose and do it better: https://github.com/nir0s/distro

We need to move away from `platform.dist()` and `platform.linux_distribution` onto either a homegrown method or nir0s/distro.

##### STEPS TO REPRODUCE
N/A

##### EXPECTED RESULTS
N/A

##### ACTUAL RESULTS
N/A

### PR DESCRIPTION
**Pull request by alikins**
##### SUMMARY

WIP fixes #35312 

Fixes #30489

Opening PR now to get CI results for some of the tests. 

##### ISSUE TYPE

 - Feature Pull Request
 - Bugfix Pull Request

##### COMPONENT NAME

lib/ansible/module_utils/distro.py

##### ANSIBLE VERSION

```
ansible 2.6.0 (platform_nir0s_distro 5483c6d09e) last updated 2018/02/09 16:38:49 (GMT -400)
  config file = /home/adrian/src/ansible/ansible.cfg
  configured module search path = [u'/home/adrian/ansible/my-modules']
  ansible python module location = /home/adrian/src/ansible/lib/ansible
  executable location = /home/adrian/src/ansible/bin/ansible
  python version = 2.7.14 (default, Jan 17 2018, 14:28:32) [GCC 7.2.1 20170915 (Red Hat 7.2.1-2)]

```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by alikins** [comment]
I failed to link this to https://github.com/ansible/ansible/issues/30489 previously.

File: lib/ansible/module_utils/distro/LICENSE [hunk 1]
```diff
@@ -0,0 +1,202 @@
+Apache License
```
Comment (mattclay) [inline lib/ansible/module_utils/distro/LICENSE]: Can the license go in the `licenses` directory and be referenced there instead?

File: lib/ansible/module_utils/distro/__init__.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+try:
+    import distro as _system_distro
+except ImportError:
+    _system_distro = None
+
+if _system_distro:
+    distro = _system_distro
+else:
+    # Our bundled copy
+    from . import _distro as distro
```
Comment (mattclay) [inline lib/ansible/module_utils/distro/__init__.py]: Our `import` sanity test currently doesn't support relative imports.

Can this be done without a relative import until that changes?

**Review by gundalow — DISMISSED** [review]
-1 from me till we've got confirmation about the license (which I'm working on)

**Comment by abadger** [comment]
@alikins I've cleaned this up a bit:
* moved thigns around
* Addressed your TODO about skipping the assert test in the test code rather than the bundled code
* Consolidates the number of commits
* Skipped our sanity tests where appropriate

I'll add a changelog entry and then this will just be waiting on legal to approve the use of Apache with GPLv3+ code.  If you'd like to take a look as the original author, feel free.

File: test/sanity/pep8/skip.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+lib/ansible/module_utils/distro/_distro.py
```
Comment (mattclay) [inline test/sanity/pep8/skip.txt:3]: Include a comment explaining why this is being skipped so someone doesn't try fixing it in the future.

**Comment by mattclay** [comment]
The pylint issue reported is a Python 2.6 compatibility issue:

```
lib/ansible/module_utils/distro/_distro.py:552:32: ansible-format-automatic-specification Format string contains automatic field numbering specification
```

That particular occurrence is a custom assert message, so it shouldn't be a problem unless the assertion fails. We can ignore the rule, but then there is a risk that we could miss other occurrences which would be a problem when updating the code.

**Comment by abadger** [comment]
We have legal permission to bundle.

@alikins, What are the ramifications of the differences between nir0s/distro and platform.dist() which the unittests you added are revealing?  Are those things that we have to work around when we hook this into facts code?  Or is it okay to use the distro values?

File: test/sanity/pep8/skip.txt (hunk 3 above)
Comment (abadger) [reply test/sanity/pep8/skip.txt:3]: Comment added

**Comment by abadger** [comment]
nitzmahone and alikins signed off on this last week on slack.  Merging for the 2.8 release.

**Pull request merged**